src/mds/MDCache.cc

```diff
@@ -1414,6 +1423,31 @@ void MDCache::adjust_subtree_after_rename(CInode *diri, CDir *olddir, bool pop)
   show_subtrees();
 }
 
+// ===================================
+// Cache Autotuning
+
+
+/*
+ * If mds_cache_autotune is set to true, then adjust cache_memory_limit using the rss usage periodically as per the mds_memory_cache_resize_interval
+ */
+void MDCache::adjust_cache_memory_limit()
+{
+  static MemoryModel mm(g_ceph_context);
+  static MemoryModel::snap last;
+  utime_t next_tuning = ceph_clock_now();
+  while(cache_autotune)
+  {
+    if (cache_resize_interval > 0 && next_tuning < ceph_clock_now()) {
+      mm.sample(&last);
+      uint64_t temp_cache = ((uint64_t) (last.get_rss()/1.25));
```

Comment:
In the general case based on quite a bit of experimentation, "RSS is 125% of the `mds_cache_memory_limit`". However, if we're autoscaling, that doesn't matter. We're just going to reduce the cache size until we're under the limit. BUT, the limit may actually be a cgroup memory limit so if we accidentally go over, the Linux oomkiller will kill the MDS. So, we need some kind of reservation as a buffer. That is the purpose of this `osd_memory_target_cgroup_limit_ratio` from https://github.com/ceph/ceph/commit/5c6b533697814af8acfd9e731a2599b2294687ef .

Please make sure to address the comment from https://github.com/ceph/ceph/pull/27027#pullrequestreview-215646551 before working on the autoscaling bit. We want the MDS to codify this `125%` formula whenever the `mds_memory_target` is set.

Reply:
Ack. The way I've implemented it( intermediary solution ), makes the memory target a hard upper bound, in the sense that in each iteration pertaining to the calculation of the mds_cache_memory_limit,  there is a check in place whether this exceeds the memory target. If yes then it skips that iteration. I can see that this would be detrimental in cases where the resource usage/RSS suddenly spikes up( slightly above the memory target) and hence due to the check in place, the cache limit remains the same. One solution that came to my mind is if we set the mds_cache_memory_limit equal to memory target whenever this happens. Thoughts?

Reply:
Here's how I am envisioning how this should work:

At startup:

````
if (cgroup_memory_limit is not infinite)
  mds_memory_target(default) = mds_memory_target_cgroup_limit_ratio * cgroup_memory_limit
mds_cache_memory_limit(default) = mds_memory_target / 1.25
````

During run-time, if `mds_cache_autotune`, in pseudo-code:

````
if (RSS > mds_memory_target)
  mds_cache_memory_limit *= 0.95
/* multiply by 0.98 to avoid thrashing the cache size changes */
else if (cache_overfull() && RSS < mds_memory_target * 0.98)
  mds_cache_memory_limit *= 1.01
````

The `cache_overfull()` check is important because we don't want to continually increase the `mds_cache_memory_limit` when there is nothing in the cache.

Reply:
Agree with:
```
else if (cache_overfull() && RSS < mds_memory_target * 0.98)
  mds_cache_memory_limit *= 1.01
```
Will add the `cache_overfull()` check too as its suboptimal to vary the mds_cache_memory_limit in every interval without checking if the cache is almost full or not.

But some doubts about the check that's happening here : 
```
if (RSS > mds_memory_target)
  mds_cache_memory_limit *= 0.95
```
Won't there be a case where initially the cache_memory_limit is way below the mds_memory_target and then due to sudden increase in the resource usage, the RSS spikes up from low to high ( above the mds_memory_target ). Shouldn't the cache_memory_limit go up in this case up till it reaches mds_memory_target?

Reply:
No. The function of `mds_memory_limit` is to set a goal RSS for the MDS. If the current RSS is greater than that goal, we need to start reducing the cache size to go back under.

One issue I didn't address in the pseudo-code above is that we need to give the MDS some time to adjust the cache via trimming and recalling capabilities from clients. We don't want to set this more than maybe once every 60 seconds and .95 may be too much a swing, .99 could be better.

Reply:
I mentioned this above, but we can't actually use RSS to decide the cache size.  I did try using RSS at first but the problem is that we can reduce the cache size with no change in RSS usage depending on wether or not the kernel can/decides to reclaim.  That can lead to extremely bad behavior.  The best we can do is look at mapped memory and try to fit it under some target.

After a discussion with Adam at Cephalocon, I'm think of adding in the ability to have to have the priority cache manager also examine total system/container memory usage and "burst" the daemon usage above the target (to some burst limit) when total RSS memory usage on the node is lower than some threshold (and scaled appropriately).  The idea here is that we can use more resources if they are available and also introduce some dynamic scaling for things like the mon where if the osd map hit rate drops too low it can be more aggressive than the OSDs about using "burst" memory to avoid monitor re-election storms.

Reply:
@markhpc I think what you really are talking about is that the allocator may not release freed memory to the kernel immediately? Perhaps instead of RSS we should be looking at `RSS() - tcmalloc_freed_but_not_released()`?